pkg/tests/end_to_end_tests/multi_tenancy_test.go

```diff
@@ -0,0 +1,580 @@
+	ts, tenants := generateSmallMultiTenantTimeseries()
+
+	withDB(t, *testDatabase, func(db *pgxpool.Pool, t testing.TB) {
+		// With valid tenants and non-tenant operations are allowed.
+		cfg := tenancy.NewSelectiveTenancyConfig(tenants[:2], true) // valid tenant-a & tenant-b.
+		mt, err := tenancy.NewAuthorizer(cfg)
+		require.NoError(t, err)
+
+		// Ingestion.
+		client, err := pgclient.NewClientWithPool(&pgclient.Config{}, 1, pgxconn.NewPgxConn(db), mt)
+		require.NoError(t, err)
+		defer client.Close()
+
+		wauth := mt.WriteAuthorizer()
+		// Ingest tenant-a.
+		request := ingstr.NewWriteRequestWithTs(copyMetrics(ts))
+		err = wauth.Process(requestWithHeaderTenant(tenants[0]), request)
+		require.NoError(t, err)
+		_, err = client.Ingest(request)
+		require.NoError(t, err)
+
+		// Ingest without tenants.
+		request = ingstr.NewWriteRequestWithTs(copyMetrics(ts))
+		err = wauth.Process(&http.Request{}, request) // Ingest without tenants.
+		require.NoError(t, err)
+		_, err = client.Ingest(request)
+		require.NoError(t, err)
+
+		ts = []prompb.TimeSeries{
+			{
+				Labels: []prompb.Label{
+					{Name: model.MetricNameLabelName, Value: "thirdMetric"},
+					{Name: "foo", Value: "bar"},
+					{Name: "common", Value: "tag"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 1, Value: 0.1},
+					{Timestamp: 2, Value: 0.2},
+					{Timestamp: 3, Value: 0.3},
+					{Timestamp: 4, Value: 0.4},
+					{Timestamp: 5, Value: 0.5},
+				},
+			},
+		}
+		request = ingstr.NewWriteRequestWithTs(copyMetrics(ts))
+		err = wauth.Process(&http.Request{}, request) // Ingest without tenants.
+		require.NoError(t, err)
+		_, err = client.Ingest(request) // Non-MT write.
+		require.NoError(t, err)
+
+		// Querying.
+		mCache := &cache.MetricNameCache{Metrics: clockcache.WithMax(cache.DefaultMetricCacheSize)}
+		lCache := clockcache.WithMax(100)
+		dbConn := pgxconn.NewPgxConn(db)
+		labelsReader := lreader.NewLabelsReader(dbConn, lCache)
+		qr := querier.NewQuerier(dbConn, mCache, labelsReader, mt.ReadAuthorizer())
+
+		// ----- query a single tenant (tenant-a) -----
+		expectedResult := []prompb.TimeSeries{
+			{
+				Labels: []prompb.Label{
+					{Name: model.MetricNameLabelName, Value: "thirdMetric"},
+					{Name: "foo", Value: "bar"},
+					{Name: "common", Value: "tag"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 2, Value: 0.2},
+					{Timestamp: 3, Value: 0.3},
+					{Timestamp: 4, Value: 0.4},
+				},
+			},
+		}
+
+		result, err := qr.Query(&prompb.Query{
+			Matchers: []*prompb.LabelMatcher{
+				{
+					Type:  prompb.LabelMatcher_EQ,
+					Name:  model.MetricNameLabelName,
+					Value: "thirdMetric",
+				},
+			},
+			StartTimestampMs: 2,
+			EndTimestampMs:   4,
+		})
+		require.NoError(t, err)
+
+		// Verifying result.
+		verifyResults(t, expectedResult, result) // Fails: expected to be 1, but shows 0 result.
+
+		// ----- query across multiple tenants (tenant-a & tenant-c) -----
```

Comment:
again would be a better test if something with tenant-c was previously ingested.